criu/bpfmap.c

```diff
@@ -0,0 +1,323 @@
+		goto err;
+	}
+	munmap(keys, bde->keys_bytes);
+	munmap(values, bde->values_bytes);
+	return 0;
+
+err:
+	munmap(keys, bde->keys_bytes);
+	munmap(values, bde->values_bytes);
+	return -1;
+}
+
+static int collect_bpfmap_data(void *obj, ProtobufCMessage *msg, struct cr_img *img)
+{
+	return do_collect_bpfmap_data(obj, msg, img, bpfmap_data_hash_table);
+}
+
+struct collect_image_info bpfmap_data_cinfo = {
+	.fd_type = CR_FD_BPFMAP_DATA,
+	.pb_type = PB_BPFMAP_DATA,
+	.priv_size = sizeof(struct bpfmap_data_rst),
+	.collect = collect_bpfmap_data,
+};
+
+int dump_one_bpfmap_data(BpfmapFileEntry *bpf, int lfd, const struct fd_parms *p)
+{
+	/*
+	 * Linux kernel patch notes for bpf_map_*_batch():
+	 *
+	 * in_batch/out_batch are opaque values use to communicate between
+	 * user/kernel space, in_batch/out_batch must be of key_size length.
+	 * To start iterating from the beginning in_batch must be null,
+	 * count is the # of key/value elements to retrieve. Note that the 'keys'
+	 * buffer must be a buffer of key_size * count size and the 'values' buffer
+	 * must be value_size * count, where value_size must be aligned to 8 bytes
+	 * by userspace if it's dealing with percpu maps. 'count' will contain the
+	 * number of keys/values successfully retrieved. Note that 'count' is an
+	 * input/output variable and it can contain a lower value after a call.
+	 * 
+	 * If there's no more entries to retrieve, ENOENT will be returned. If error
+	 * is ENOENT, count might be > 0 in case it copied some values but there were
+	 * no more entries to retrieve.
+	 * 
+	 * Note that if the return code is an error and not -EFAULT,
+	 * count indicates the number of elements successfully processed.
+	 */
+
+	struct cr_img *img;
+	uint32_t key_size, value_size, max_entries, count;
+	void *keys = NULL, *values = NULL;
+	void *in_batch = NULL, *out_batch = NULL;
+	BpfmapDataEntry bde = BPFMAP_DATA_ENTRY__INIT;
+	DECLARE_LIBBPF_OPTS(bpf_map_batch_opts, opts,
+		.elem_flags = 0,
+		.flags = 0,
+	);
+	int ret;
+
+	key_size = bpf->key_size;
+	value_size = bpf->value_size;
+	max_entries = bpf->max_entries;
+	count = max_entries;
+
+	keys = mmap(NULL, key_size * max_entries, PROT_READ | PROT_WRITE,
+				MAP_SHARED | MAP_ANONYMOUS, 0, 0); 
+	if(keys == MAP_FAILED) {
+		pr_perror("Can't map memory for BPF map keys\n");
+		goto err;
+	}
+
+	values = mmap(NULL, value_size * max_entries, PROT_READ | PROT_WRITE,
+				MAP_SHARED | MAP_ANONYMOUS, 0, 0);
+	if(values == MAP_FAILED) {
+		pr_perror("Can't map memory for BPF map values\n");
+		goto err;
+	}
+
+	out_batch = mmap(NULL, key_size, PROT_READ | PROT_WRITE,
+				MAP_SHARED | MAP_ANONYMOUS, 0, 0); 
+	if(out_batch == MAP_FAILED) {
+		pr_perror("Can't map memory for BPF map out_batch\n");
+		goto err;
+	}
+
+	ret = bpf_map_lookup_batch(lfd, in_batch, out_batch, keys, values, &count, &opts);
+	if(ret && errno != ENOENT) {
+		pr_perror("Can't perform a batch lookup on BPF map\n");
+		goto err;
+	}
+
```

Comment:
shouldn't we have a check about `count == 0`?

Reply:
If count == 0 after the lookup, I store it as 0 in the protobuf image (without any data).
During restore, I check to see whether the count is 0 and if this is the case, return without trying to restore data:
https://github.com/checkpoint-restore/criu/pull/1044/files#diff-72d5d16cff7b2f3309c3fd148760d60aR82-R85

Would you like me to change the logic such that it checks for count == 0 during the dump phase instead of restore? If so, I would have to return from the function only after freeing the memory allocated above. I think the cleanest way to do this would be by adding a goto label 'pass' here:
https://github.com/checkpoint-restore/criu/pull/1044/files#diff-72d5d16cff7b2f3309c3fd148760d60aR216
The code would then become:

`ret = bpf_map_lookup_batch(lfd, in_batch, out_batch, keys, values, &count, &opts);

	if (ret && errno != ENOENT) {
		pr_perror("Can't perform a batch lookup on BPF map");
		goto err;
	}
        if (count == 0)
                goto pass;`

Please let me know which approach you think would be better.

Thank you.